### PR TITLE
Bootstrap module

### PR DESCRIPTION
**Pull request by chamerling**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,6 +20,14 @@
     "url": "https://github.com/OpenPaaS-Suite/openpaas-auth-client/issues"
   },
   "homepage": "https://github.com/OpenPaaS-Suite/openpaas-auth-client#readme",
+  "dependencies": {
+    "@babel/core": "^7.10.5",
+    "@babel/preset-env": "^7.10.4",
+    "babel-loader": "^8.1.0",
+    "oidc-client": "^1.10.1",
+    "webpack": "^4.44.1",
+    "webpack-cli": "^3.3.12"
```
Comment (tuanlc) [inline package.json]: `babel*` & `webpack*` should be `devDependencies`?

File: src/basic/index.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/**
+ * Basic Auth plugin.
+ */
+class BasicAuth {
+  constructor({ loginPath = '/login', logoutPath = '/logout', fetchUser = null } = {}) {
```
Comment (tuanlc) [inline src/basic/index.js:5]:
```suggestion
  constructor({ loginPath, logoutPath, fetchUserl } = { loginPath = '/login', logoutPath = '/logout' }) {
```

File: src/basic/index.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    this.logoutPath = logoutPath;
+    this.fetchUser = fetchUser;
+
+    if (!this.fetchUser) {
+      throw new Error('fetchUser is required');
+    }
+  }
+
+  init() {
+    return this.fetchUser()
```
Comment (tuanlc) [inline src/basic/index.js:16]: Should we require `fetchUser` is a function?

File: src/basic/index.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    return Promise.resolve();
+  }
+
+  isLoggedIn() {
+    return !!this.user;
+  }
+
+  logout() {
+    window.location = this.options.logoutPath;
```
Comment (piavgh) [inline src/basic/index.js]: where does the `.options` come from?

File: src/basic/index.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      .then(user => {
+        this.user = user;
+
+        return this.user;
+      })
+      .catch(() => false);
+  }
+
+  login() {
+    window.location = `${this.loginPath}?continue=${window.location.hash}`;
```
Comment (piavgh) [inline src/basic/index.js:26]: if we use `window.location.hash` like this, there is a problem that I met here: https://github.com/OpenPaaS-Suite/esn/issues/16#issuecomment-672648286

File: package.json [hunk 6]
```diff
@@ -5,7 +5,10 @@
   "main": "src/index.js",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "lint": "eslint . --ext '.js' --cache"
+    "lint": "eslint . --ext '.js' --cache",
+    "build:dev": "webpack --mode development",
+    "build:prod": "webpack --mode production",
+    "postinstall": "webpack --mode production"
```
Comment (piavgh) [inline package.json]: Why do we need this `postinstall` script?

File: src/basic/index.js (hunk 2 above)
Comment (MichaelBailly) [reply src/basic/index.js:5]: The former is easier to read than your proposal Tuan, IMHO

File: src/basic/index.js (hunk 3 above)
Comment (chamerling) [reply src/basic/index.js:16]: yes!

File: src/basic/index.js (hunk 5 above)
Comment (chamerling) [reply src/basic/index.js:26]: let's complete https://github.com/OpenPaaS-Suite/esn/issues/16#issuecomment-672648286 first then

File: src/basic/index.js (hunk 4 above)
Comment (chamerling) [reply src/basic/index.js]: good catch

File: package.json (hunk 1 above)
Comment (chamerling) [reply package.json]: moved, will see if install works well in this case

File: package.json (hunk 6 above)
Comment (chamerling) [reply package.json]: Removed

File: src/oidc/index.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        return this._onSignIn();
+      });
+  }
+
+  _onSignIn() {
+    this.options.onSignInComplete &&
+    typeof this.options.onSignInComplete === 'function' &&
+    this.options.onSignInComplete({ headers: { Authorization: this.strategy.getAuthorizationHeaderValue()}});
```
Comment (itswillta) [inline src/oidc/index.js]:
```suggestion
    typeof this.options.onSignInComplete === 'function' &&
    this.options.onSignInComplete({ headers: { Authorization: this.strategy.getAuthorizationHeaderValue()}});
```

File: src/oidc/index.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    this.strategy = new OIDCStrategy(options);
+    this.callbackURLPattern = options.callbackURLPattern || DEFAULT_URL_PATTERN;
+  }
+
+  init() {
+    // URL may be updated before now and the strategy.init promise resolution...
+    const currentUrl = window.location.href;
+
+    return this.strategy.init()
+      .then(() => {
```
Comment (itswillta) [inline src/oidc/index.js:17]: We're using Babel, so can we use async await here?

Comment (chamerling) [reply src/oidc/index.js:17]: This is a global decision we did not make for now

**Review by tuanlc — APPROVED** [review]

**Review by itswillta — APPROVED** [review]

**Pull request merged**